---
title: "Analysis"
subtitle: Andreas Vija
output:
  html_document:
    df_print: paged
  pdf: default
  pdf_document: default
editor_options:
  chunk_output_type: console
---

# Setup

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = FALSE, warning = FALSE)
```

```{r}
wd = "~/cage/analysis/"
setwd(wd)

library("GenomicFeatures")
library("dplyr")
library("ggplot2")
library("readr")
library("stringr")
library("reshape2")
library("tidyr")

set.seed(123)
```

# Load in

Starting from now, code requires qtlmap output for CAGE

```{r}
col_names = c("gene", "top_phenotype", "total_phenotypes", "variants_tested", "top_variant", "top_variant_chr", "top_variant_start", "p_nominal", "beta", "p_adjusted", "p_adjusted_beta")
col_types = "cciiccidddd"

bwa_results = read_delim("bwa.permuted.txt", delim="\t", col_types=col_types, col_names=col_names, skip=1)
```

Starting from now, code also requires regular txrevise output

```{r}
upstream1 = GenomicFeatures::makeTxDbFromGFF(
  "txrevise/scripts/processed/Homo_sapiens.GRCh38.96_regular/txrevise_regular.grp_1.upstream.gff3")
upstream2 = GenomicFeatures::makeTxDbFromGFF(
  "txrevise/scripts/processed/Homo_sapiens.GRCh38.96_regular/txrevise_regular.grp_2.upstream.gff3")
genes1 = AnnotationDbi::select(upstream1, keys=keys(upstream1, "GENEID"), keytype="GENEID", columns="GENEID")
genes2 = AnnotationDbi::select(upstream2, keys=keys(upstream2, "GENEID"), keytype="GENEID", columns="GENEID")
genes1 = unlist(genes1, use.names=FALSE)
genes2 = unlist(genes2, use.names=FALSE)

genes = union(sapply(genes1, fun), sapply(genes2, fun))
```

```{r}
common_genes = intersect(bwa_results$gene, genes)
saveRDS(common_genes, "common_genes.rds")
```

Starting from now, code also requires qtlmap output for txrevise and txrevise + new annotations

```{r}
txrevise_results = read_delim("txrev.permuted.txt", delim="\t", col_types=col_types, col_names=col_names, skip=1)
annots_results = read_delim("txrev_annots_25.permuted.txt", delim="\t", col_types=col_types, col_names=col_names, skip=1)

bwa_results = bwa_results %>%
  mutate(kind="upstream") # for homogeneity's sake
fun = function(x, n=1) {
  return(strsplit(x, ".", fixed=TRUE)[[1]][n])
}
txrevise_results = txrevise_results %>%
  mutate(kind=sapply(gene, fun, n=2), gene=sapply(gene, fun, n=1)) %>%
  filter(kind=="upstream")
annots_results = annots_results %>%
  mutate(kind=sapply(gene, fun, n=2), gene=sapply(gene, fun, n=1)) %>%
  filter(kind=="upstream")
```

```{r}
bwa_results_common = bwa_results[bwa_results$gene %in% common_genes,]
txrevise_results_common = txrevise_results[txrevise_results$gene %in% common_genes,]
annots_results_common = annots_results[annots_results$gene %in% common_genes,]
```

```{r}
bwa_results_common$p_adjusted_beta_adjusted = p.adjust(bwa_results_common$p_adjusted_beta, method="fdr")
txrevise_results_common$p_adjusted_beta_adjusted = p.adjust(txrevise_results_common$p_adjusted_beta, method="fdr")
annots_results_common$p_adjusted_beta_adjusted = p.adjust(annots_results_common$p_adjusted_beta, method="fdr")

# just to keep a consistent shape
bwa_results$p_adjusted_beta_adjusted = p.adjust(bwa_results$p_adjusted_beta, method="fdr")
txrevise_results$p_adjusted_beta_adjusted = p.adjust(txrevise_results$p_adjusted_beta, method="fdr")
annots_results$p_adjusted_beta_adjusted = p.adjust(annots_results$p_adjusted_beta, method="fdr")
```

```{r}
fill_genes <- function(x, common=FALSE) {
  gene_source = genes
  if (common) {
    gene_source = common_genes
  }
  old_genes = unlist(x$gene, use.names=FALSE)
  new_genes = setdiff(gene_source, old_genes)

  additional = data_frame(gene=new_genes, top_phenotype="_", total_phenotypes=1, variants_tested=1, top_variant="_", 
                          top_variant_chr="_", top_variant_start=1, p_nominal=0.99, beta=1, p_adjusted=0.99, 
                          p_adjusted_beta=0.99, p_adjusted_beta_adjusted=0.99, kind="upstream")
  return(rbind(x, additional))
}

txrevise_results = fill_genes(txrevise_results)
txrevise_results_common = fill_genes(txrevise_results_common, common=TRUE)
annots_results = fill_genes(annots_results)
annots_results_common = fill_genes(annots_results_common, common=TRUE)
```

```{r}
ggplot(mapping=aes(x=bwa_results$p_adjusted_beta_adjusted)) +
  geom_histogram()
sum(bwa_results$p_adjusted_beta_adjusted < 0.05)

ggplot(mapping=aes(x=txrevise_results$p_adjusted_beta_adjusted)) +
  geom_histogram()
sum(txrevise_results$p_adjusted_beta_adjusted < 0.05)

ggplot(mapping=aes(x=annots_results$p_adjusted_beta_adjusted)) +
  geom_histogram()
sum(annots_results$p_adjusted_beta_adjusted < 0.05)

#ggplot(mapping=aes(x=abs(bwa_results$variant_distance[bwa_results$p_adjusted_beta_adjusted < 0.05]))) +
#  geom_histogram()
```

```{r}
bwa_tops = bwa_results_common %>%
  filter(p_adjusted_beta_adjusted < 0.05) %>%
  arrange(p_adjusted_beta_adjusted)

tx_tops = txrevise_results_common %>%
  filter(p_adjusted_beta_adjusted < 0.05) %>%
  arrange(p_adjusted_beta_adjusted)

ann_tops = annots_results_common %>%
  filter(p_adjusted_beta_adjusted < 0.05) %>%
  arrange(p_adjusted_beta_adjusted)
```

# Analysis

```{r}
dim(bwa_results)[1] # BWA total: 19 113
sum(bwa_results$total_phenotypes) # BWA promotor total: 91 521 (pre-qtlmap number bigger cause genes with not enough variants discarded?)
dim(txrevise_results)[1] # txrevise total: 20 176
dim(annots_results)[1] # annots total: 20 176
length(intersect(bwa_results$gene, txrevise_results$gene)) # common total: 15 690
length(intersect(bwa_results$gene, annots_results$gene)) # same

sum(txrevise_results$total_phenotypes) # 79 963
sum(annots_results$total_phenotypes) # 83 064
sum(annots_results$total_phenotypes) - sum(txrevise_results$total_phenotypes) # 3301, expected N=25 -> 38 730 annotations added

dim(bwa_tops)[1] # BWA tops: 1217
dim(tx_tops)[1] # txrevise tops: 1086
dim(ann_tops)[1] # annots tops: [10]  (+) [25] 1068 (-18)

removed = setdiff(tx_tops$gene, ann_tops$gene)
length(removed) # annotations removed: [10]  [25] 83
added = setdiff(ann_tops$gene, tx_tops$gene)
length(added) # annotations added: [10]  [25] 65

length(intersect(bwa_tops$gene, tx_tops$gene)) # common both tops: 305
length(intersect(bwa_tops$gene, ann_tops$gene)) # common after tops: [10]  (+) [25] 304 (-1)

length(intersect(bwa_tops$top_variant, tx_tops$top_variant)) # 41
length(intersect(bwa_tops$top_variant, ann_tops$top_variant)) # [10]  [25] 43
```


```{r}
top_variants = rbind(bwa_tops[c("top_variant")], tx_tops[c("top_variant")], ann_tops[c("top_variant")]) %>%
  unique()
write_delim(top_variants, "variants.txt")

to_visualize = bwa_tops[! bwa_tops$gene %in% tx_tops$gene,] %>% #912
  select(gene, top_variant)
write_tsv(to_visualize, "cage_over_tx.tsv")

to_visualize = tx_tops[! tx_tops$gene %in% bwa_tops$gene,] %>% #781
  select(gene, top_variant)
write_tsv(to_visualize, "tx_over_cage.tsv")
```

```{r}
# which genes and variants were newly significant after new annots but not with cage (false positives?)

interesting1 = bwa_results[bwa_results$gene %in% added,c("gene", "p_adjusted_beta_adjusted")]
colnames(interesting1) = c("gene", "p_cage")
interesting2 = ann_tops[ann_tops$gene %in% added, c("gene", "p_adjusted_beta_adjusted")]
colnames(interesting2) = c("gene", "p_ann")

interesting = merge(interesting1, interesting2) %>%
  filter(p_cage > 0.05, p_ann < 0.05)

to_visualize = ann_tops[ann_tops$gene %in% interesting$gene,] %>%
  select(gene, top_variant)
write_tsv(to_visualize, "ann_over_cage.tsv")
```

```{r}
getgrouping = function(x) {
  if (x %in% added) {return("Muutus oluliseks")}
  if (x %in% removed) {return("Kaotas olulisuse")}
  return("Ei muutunud")
}

comparison = merge(select(bwa_results, gene, p_adjusted_beta_adjusted),
                   select(txrevise_results, gene, p_adjusted_beta_adjusted),
                   by="gene")
colnames(comparison) = c("gene", "bwa_p", "tx_p")
comparison$grouping = as.factor(sapply(comparison$gene, getgrouping))
comparison = arrange(comparison, grouping)

fig1 = ggplot(comparison, aes(x=-log10(tx_p), y=-log10(bwa_p), color=grouping)) +
  labs(x="txrevise -log10(p)", y="CAGE -log10(p)", color="Olulisus") +
  geom_smooth(aes(x=-log10(tx_p), y=-log10(bwa_p)), inherit.aes=FALSE, method="lm")


comparison = merge(select(bwa_results, gene, p_adjusted_beta_adjusted),
                   select(annots_results, gene, p_adjusted_beta_adjusted),
                   by="gene")
colnames(comparison) = c("gene", "bwa_p", "ann_p")
comparison$grouping = as.factor(sapply(comparison$gene, getgrouping))
comparison = arrange(comparison, grouping)

fig2 = ggplot(comparison, aes(x=-log10(ann_p), y=-log10(bwa_p), color=grouping)) +
  labs(x="TÃ¤iendatud txrevise -log10(p)", y="CAGE -log10(p)", color="Olulisus") +
  geom_smooth(aes(x=-log10(ann_p), y=-log10(bwa_p)), inherit.aes=FALSE, method="lm")


comparison = merge(select(annots_results, gene, p_adjusted_beta_adjusted),
                   select(txrevise_results, gene, p_adjusted_beta_adjusted),
                   by="gene")
colnames(comparison) = c("gene", "ann_p", "tx_p")
comparison$grouping = as.factor(sapply(comparison$gene, getgrouping))
comparison = arrange(comparison, grouping)

fig3 = ggplot(comparison, aes(x=-log10(tx_p), y=-log10(ann_p), color=grouping)) +
  labs(x="txrevise -log10(p)", y="TÃ¤iendatud txrevise -log10(p)", color="Olulisus") +
  geom_smooth(aes(x=-log10(tx_p), y=-log10(ann_p)), inherit.aes=FALSE, method="lm")


for (fig in list(fig1, fig2, fig3)) {
  print(fig +
          geom_segment(x=0, y=0, xend=120, yend=120, color="black") +
          geom_point(aes(color=grouping), alpha=0.4) +
          scale_color_manual(values=c("dark gray", "red", "green")) +
          xlim(0,50) + ylim(0,50) +
          theme(legend.position="top"))
}
```

```{r}
sample_mapping = read_tsv("../qtlmap_prep/sampleMetadata.tsv") %>%
  select(sample_id, genotype_id) %>%
  mutate(source="cage")
sample_mapping_ = read_tsv("GEUVADIS_EUR.tsv") %>%
  select(sample_id, genotype_id) %>%
  mutate(source="txrevise")
sample_mapping = rbind(sample_mapping, sample_mapping_)

genotypes = read_tsv("variantinfo.vcf")[c(3,10:163)] %>%
  melt(id.vars=c("ID")) %>%
  mutate(source="cage")
genotypes_ = read_tsv("variantinfo_geuvadis.vcf")[c(3,10:454)] %>%
  melt(id.vars=c("ID")) %>%
  mutate(source="txrevise")
genotypes = rbind(genotypes, genotypes_)

names(genotypes) = c("variant", "genotype_id", "alleles", "source")
genotypes = genotypes %>%
  merge(sample_mapping) %>%
  select(-genotype_id, -source)

mapping = list(0, 1, 1, 2)
names(mapping) = c("0|0", "0|1", "1|0", "1|1")
genotypes$alleles = mapping[genotypes$alleles]

counts_bwa = read_tsv("../qtlmap_prep/countMatrix.tsv") %>%
  melt(id.vars=c("phenotype_id"), variable.name=c("sample_id"), value.name="proportion")
#counts_tx = read_tsv("/countMatrix_txrevise.tsv") %>% # 2.3G raw
#  melt(id.vars=c("phenotype_id"), variable.name=c("sample_id"), value.name="proportion")

#dataset, counts file, promoter, variant
for (info in list(
  list(bwa_tops, 1, counts_bwa),
  list(bwa_tops, 2, counts_bwa)
)) {
  dataset = info[[1]]
  n = info[[2]]
  counts = info[[3]]

  this_promoter = dataset$top_phenotype[n]
  this_variant = dataset$top_variant[n]

  counts_subset = filter(counts, phenotype_id==this_promoter) %>%
    select(sample_id, proportion)
  genotypes_subset = filter(genotypes, variant==this_variant) %>%
    select(sample_id, alleles)
  to_plot = merge(counts_subset, genotypes_subset)
  to_plot$alleles = as.numeric(to_plot$alleles)

  coef = coef(lm(proportion ~ alleles, data=to_plot))
  p = ggplot(to_plot, aes(x=alleles, y=proportion, group=as.factor(alleles))) +
        geom_jitter(width=0.1, color="black", alpha=0.7) +
        geom_boxplot(alpha=0, lwd=1, color="black") +
        scale_x_continuous(breaks=c(0,1,2)) +
        geom_abline(intercept=coef[1], slope=coef[2], lwd=1) +
        labs(x="Aternatiivsete alleelide arv", y="Promootori normaliseeritud osakaalud")
  print(p)
}
```
