---
title: "Analysis"
subtitle: Andreas Vija
output:
  html_document:
    df_print: paged
  pdf: default
  pdf_document: default
editor_options:
  chunk_output_type: console
---

# Setup

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = FALSE, warning = FALSE)

wd = "~/cage/analysis/"
setwd(wd)

library("GenomicFeatures")
library("dplyr")
library("ggplot2")
library("readr")
library("stringr")
library("reshape2")
library("tidyr")
library("mixtools")
library("stats")

set.seed(123)
```

# Load in

```{r}
col_names = c("gene", "top_phenotype", "total_phenotypes", "variants_tested", "top_variant", "top_variant_chr", "top_variant_start", "p_nominal", "beta", "p_adjusted", "p_adjusted_beta")
col_types = "cciiccidddd"

bwa_results = read_delim("bwa.permuted.txt", delim="\t", col_types=col_types, col_names=col_names, skip=1)
txrevise25_results = read_delim("txrev-25.permuted.txt", delim="\t", col_types=col_types, col_names=col_names, skip=1)
annots_results = read_delim("txrev_annots_20.permuted.txt", delim="\t", col_types=col_types, col_names=col_names, skip=1)

bwa_results = bwa_results %>%
  mutate(kind="upstream") # for homogeneity's sake
fun = function(x, n=1) {
  return(strsplit(x, ".", fixed=TRUE)[[1]][n])
}
txrevise25_results = txrevise25_results %>%
  mutate(kind=sapply(gene, fun, n=2), gene=sapply(gene, fun, n=1)) %>%
  filter(kind=="upstream")
annots_results = annots_results %>%
  mutate(kind=sapply(gene, fun, n=2), gene=sapply(gene, fun, n=1)) %>%
  filter(kind=="upstream")

upstream1 = GenomicFeatures::makeTxDbFromGFF(
  "txrevise/scripts/processed/Homo_sapiens.GRCh38.96_regular/txrevise_regular.grp_1.upstream.gff3")
upstream2 = GenomicFeatures::makeTxDbFromGFF(
  "txrevise/scripts/processed/Homo_sapiens.GRCh38.96_regular/txrevise_regular.grp_2.upstream.gff3")
genes1 = AnnotationDbi::select(upstream1, keys=keys(upstream1, "GENEID"), keytype="GENEID", columns="GENEID")
genes2 = AnnotationDbi::select(upstream2, keys=keys(upstream2, "GENEID"), keytype="GENEID", columns="GENEID")
genes1 = unlist(genes1, use.names=FALSE)
genes2 = unlist(genes2, use.names=FALSE)
genes = union(sapply(genes1, fun), sapply(genes2, fun))

grp_1_exon_names = names(GenomicFeatures::exonsBy(upstream1, by = "tx", use.names = TRUE))
grp_2_exon_names = names(GenomicFeatures::exonsBy(upstream2, by = "tx", use.names = TRUE))

remove(upstream1, upstream2, genes1, genes2)

beginning_common_genes = intersect(bwa_results$gene, genes)
common_genes = beginning_common_genes
# common genetic variants already only used in qtlmap, but bwa_results' 18k > common genes' 15k
```

# Filter results

```{r}
bwa_results_common = bwa_results[bwa_results$gene %in% common_genes,]
txrevise25_results_common = txrevise25_results[txrevise25_results$gene %in% common_genes,]
annots_results_common = annots_results[annots_results$gene %in% common_genes,]

bwa_results_common$p_adjusted_beta_adjusted = p.adjust(bwa_results_common$p_adjusted_beta, method="fdr")
txrevise25_results_common$p_adjusted_beta_adjusted = p.adjust(txrevise25_results_common$p_adjusted_beta, method="fdr")
annots_results_common$p_adjusted_beta_adjusted = p.adjust(annots_results_common$p_adjusted_beta, method="fdr")

# just to keep a consistent shape
bwa_results$p_adjusted_beta_adjusted = p.adjust(bwa_results$p_adjusted_beta, method="fdr")
txrevise25_results$p_adjusted_beta_adjusted = p.adjust(txrevise25_results$p_adjusted_beta, method="fdr")
annots_results$p_adjusted_beta_adjusted = p.adjust(annots_results$p_adjusted_beta, method="fdr")


fill_genes <- function(x, common=FALSE) {
  gene_source = genes
  if (common) {
    gene_source = common_genes
  }
  old_genes = unlist(x$gene, use.names=FALSE)
  new_genes = setdiff(gene_source, old_genes)

  additional = data_frame(gene=new_genes, top_phenotype="_", total_phenotypes=0, variants_tested=0, top_variant="_",
                          top_variant_chr="_", top_variant_start=1, p_nominal=0.99, beta=1, p_adjusted=0.99,
                          p_adjusted_beta=0.99, p_adjusted_beta_adjusted=0.99, kind="upstream")
  return(rbind(x, additional))
}

txrevise25_results = fill_genes(txrevise25_results)
txrevise25_results_common = fill_genes(txrevise25_results_common, common=TRUE)
annots_results = fill_genes(annots_results)
annots_results_common = fill_genes(annots_results_common, common=TRUE)


#ggplot(mapping=aes(x=bwa_results$p_adjusted_beta_adjusted)) +
#  geom_histogram()

#ggplot(mapping=aes(x=txrevise25_results$p_adjusted_beta_adjusted)) +
#  geom_histogram()

#ggplot(mapping=aes(x=annots_results$p_adjusted_beta_adjusted)) +
#  geom_histogram()

#ggplot(mapping=aes(x=abs(bwa_results$variant_distance[bwa_results$p_adjusted_beta_adjusted < 0.05]))) +
#  geom_histogram()


bwa_tops = bwa_results_common %>%
  filter(p_adjusted_beta_adjusted < 0.05) %>%
  arrange(p_adjusted_beta_adjusted)

tx25_tops = txrevise25_results_common %>%
  filter(p_adjusted_beta_adjusted < 0.05) %>%
  arrange(p_adjusted_beta_adjusted)

ann_tops = annots_results_common %>%
  filter(p_adjusted_beta_adjusted < 0.05) %>%
  arrange(p_adjusted_beta_adjusted)
```

# Analysis

```{r}
dim(bwa_results)[1] # BWA total: 18 432
sum(bwa_results$total_phenotypes) # BWA promotor total: 89 594 (pre-qtlmap number bigger cause genes with not enough variants discarded?)
dim(txrevise25_results)[1] # 20 176
dim(annots_results)[1] # same
length(intersect(bwa_results$gene, txrevise25_results$gene)) # 15 275
length(intersect(bwa_results$gene, annots_results$gene)) # same

sum(txrevise25_results_common$total_phenotypes) # 60 215
sum(annots_results_common$total_phenotypes) # 94 385
sum(annots_results_common$total_phenotypes) - sum(txrevise25_results_common$total_phenotypes) # 34 170

dim(bwa_tops)[1] # BWA tops: 1145
dim(tx25_tops)[1] # txrevise25 tops: 979
dim(ann_tops)[1] # annots tops: 1287
(dim(ann_tops)[1] - dim(tx25_tops)[1]) / dim(tx25_tops)[1] # 31.5%

length(intersect(bwa_tops$gene, tx25_tops$gene)) # 307
length(intersect(bwa_tops$gene, ann_tops$gene)) # 397
length(intersect(tx25_tops$gene, ann_tops$gene)) # 806

length(intersect(bwa_tops$top_variant, tx25_tops$top_variant)) # 52
length(intersect(bwa_tops$top_variant, ann_tops$top_variant)) # 56

length(grp_1_exon_names) # 61 453
length(grp_2_exon_names) # 48 394
length(intersect(grp_1_exon_names, grp_2_exon_names)) # 0
length(union(grp_1_exon_names, grp_2_exon_names)) # 109 847
new_transcripts = readRDS("txrevise/scripts/processed/Homo_sapiens.GRCh38.96_CAGE-20/new_transcripts_20.rds")
length(new_transcripts) # 77 869
```

# To visualize

```{r}
top_variants = rbind(bwa_tops[c("top_variant")], tx25_tops[c("top_variant")], ann_tops[c("top_variant")]) %>%
  unique() # 2774
#write_delim(top_variants, "variants.txt")
remove(top_variants)

cage_over_tx = bwa_tops[! bwa_tops$gene %in% tx25_tops$gene,] %>% # 838
  select(gene, top_variant)
#write_tsv(cage_over_tx, "cage_over_tx.tsv")

tx_and_cage = bwa_tops[bwa_tops$gene %in% tx25_tops$gene,] %>% # 307
  select(gene, top_variant)
#write_tsv(tx_and_cage, "tx_and_cage.tsv")

tx_over_cage = tx25_tops[! tx25_tops$gene %in% bwa_tops$gene,] %>% # 672
  select(gene, top_variant)
#write_tsv(tx_over_cage, "tx_over_cage.tsv")

# which genes and variants were newly significant after new annots but not with cage (false positives?)
added = setdiff(ann_tops$gene, tx25_tops$gene)
interesting1 = bwa_results[bwa_results$gene %in% added, c("gene", "p_adjusted_beta_adjusted")]
colnames(interesting1) = c("gene", "p_cage")
interesting2 = ann_tops[ann_tops$gene %in% added, c("gene", "p_adjusted_beta_adjusted")]
colnames(interesting2) = c("gene", "p_ann")
remove(added)

interesting = merge(interesting1, interesting2) %>%
  filter(p_cage > 0.05, p_ann < 0.05)

ann_over_cage = ann_tops[ann_tops$gene %in% interesting$gene,] %>% # 353
  select(gene, top_variant)
remove(interesting, interesting1, interesting2)
#write_tsv(ann_over_cage, "ann_over_cage.tsv")
```

# Optimizing annotations' N

```{r}
compare_tx_tops = tx25_tops

transcript_count = c()
common_transcript_count = c()
affected_gene_count = c()
affected_common_gene_count = c()

added_genes = c()
removed_genes = c()
net_increase = c()

added_in_bwa_tops = c()
frac_added_in_bwa_tops = c()

Ns = c(5, 10, 15, 20, 25, 50, 75, 100, 150, 200)
for (N in Ns) {
  new_transcripts = readRDS(paste0("txrevise/scripts/processed/Homo_sapiens.GRCh38.96_CAGE-", N, "/new_transcripts_", N, ".rds"))
  transcript_count = c(transcript_count, length(new_transcripts))
  common_transcript_count = c(common_transcript_count, sum(sapply(new_transcripts@partitioning@NAMES, fun, 1) %in% common_genes))

  affected_genes = readRDS(paste0("txrevise/scripts/processed/Homo_sapiens.GRCh38.96_CAGE-", N, "/new_transcript_genes_", N, ".rds"))
  affected_gene_count = c(affected_gene_count, length(affected_genes))
  affected_common_gene_count = c(affected_common_gene_count, length(intersect(affected_genes, common_genes)))

  trial_results = read_delim(paste0("txrev_annots_", N, ".permuted.txt"), delim="\t", col_types=col_types, col_names=col_names, skip=1)
  trial_results = trial_results %>%
    mutate(kind=sapply(gene, fun, n=2), gene=sapply(gene, fun, n=1)) %>%
    filter(kind=="upstream")
  trial_results_common = trial_results[trial_results$gene %in% common_genes,]
  trial_results_common$p_adjusted_beta_adjusted = p.adjust(trial_results_common$p_adjusted_beta, method="fdr")
  trial_results$p_adjusted_beta_adjusted = p.adjust(trial_results$p_adjusted_beta, method="fdr")
  trial_results = fill_genes(trial_results)
  trial_results_common = fill_genes(trial_results_common, common=TRUE)
  trial_tops = trial_results_common %>%
    filter(p_adjusted_beta_adjusted < 0.05) %>%
    arrange(p_adjusted_beta_adjusted)

  added_this = setdiff(trial_tops$gene, compare_tx_tops$gene)
  added_genes = c(added_genes, length(added_this))
  removed_genes = c(removed_genes, length(setdiff(compare_tx_tops$gene, trial_tops$gene)))
  net_increase = c(net_increase, dim(trial_tops)[1] - dim(compare_tx_tops)[1])

  added_in_bwa_tops_this = intersect(added_this, bwa_tops$gene)
  added_in_bwa_tops = c(added_in_bwa_tops, length(added_in_bwa_tops_this))
  frac_added_in_bwa_tops = c(frac_added_in_bwa_tops, length(added_in_bwa_tops_this) / length(added_this))
}

metrics = tibble(N=Ns,
                 transcript_count = transcript_count,
                 common_transcript_count = common_transcript_count,
                 affected_gene_count = affected_gene_count,
                 affected_common_gene_count = affected_common_gene_count)

melted_metrics = melt(metrics, id.vars="N", variable.name="metric")
ggplot(melted_metrics, aes(N, value)) +
  geom_line() +
  geom_point() +
  facet_grid(metric ~ ., scales="free_y") +
  scale_x_continuous(breaks=Ns)

metrics = tibble(N=Ns,
                 `Net increase` = net_increase,
                 #added_genes = added_genes,
                 `Removed genes` = removed_genes,
                 #added_in_bwa_tops = added_in_bwa_tops,
                 `Shared with` = frac_added_in_bwa_tops)

melted_metrics = melt(metrics, id.vars="N", variable.name="metric")
ggplot(melted_metrics, aes(N, value)) +
  geom_line() +
  geom_point() +
  facet_grid(metric ~ ., scales="free_y") +
  scale_x_continuous(breaks=Ns)
```

# Calculate expression

```{r}
counts = read_tsv("txrev-25_merged_gene_counts.txt") %>%
  tidyr::separate(phenotype_id, c("gene", "second", "third")) %>%
  filter(is.na(third)) %>%
  select(-second, -third)

gene_metadata = read_tsv("../qtlmap_prep/transcript_usage_Ensembl_96_phenotype_metadata.tsv", col_types="ccccciicccci") %>%
  mutate(gene_length=gene_end-gene_start) %>%
  select(gene_id, gene_length) %>%
  unique()

length(gene_metadata$gene_id) - length(unique(gene_metadata$gene_id)) # 0
length(intersect(gene_metadata$gene_id, counts$gene)) / length(counts$gene) # 99.3%
not_found = setdiff(counts$gene, gene_metadata$gene_id)
counts = counts %>% filter(!(gene %in% not_found))

rownames(gene_metadata) = gene_metadata$gene_id
gene_lengths = gene_metadata[counts$gene,]$gene_length

# divide by lengths of genes in kilobases, get reads per kilobase
counts[2:length(counts)] = counts[2:length(counts)] / (gene_lengths / 1000)
# scale it to per million, get TPM
scaling_factors = 1000000 / colSums(counts[2:length(counts)])
counts[2:length(counts)] = sweep(counts[2:length(counts)], 2, scaling_factors, "*")

counts = counts %>%
  melt(id.vars=c("gene"), variable.name="sample", value.name="TPM")
dim(filter(counts, TPM == 0))[1] / dim(counts)[1] # 57.1%
```

# Filter expression

```{r}
cage_only_top_genes = cage_over_tx$gene
tx_only_top_genes = tx_over_cage$gene
both_top_genes = tx_and_cage$gene

counts[,"top_in"] = "None"
counts[counts$gene %in% cage_only_top_genes,"top_in"] = "CAGE"
counts[counts$gene %in% tx_only_top_genes,"top_in"] = "txrevise"
counts[counts$gene %in% both_top_genes,"top_in"] = "Both"

counts_common = counts %>%
  filter(gene %in% common_genes)
dim(filter(counts_common, TPM == 0))[1] / dim(counts_common)[1] # 12.4% at 0 TPM, 0.2% at 1TPM

counts_tops = counts %>%
  filter(top_in != "None")
dim(filter(counts_tops, TPM == 0))[1] / dim(counts_tops)[1] # 0.1% at 1 TPM


dim(filter(counts_tops, TPM == 0, top_in == "CAGE"))[1] / dim(filter(counts_tops, top_in == "CAGE"))[1] # <0.1%
dim(filter(counts_tops, TPM == 0, top_in == "txrevise"))[1] / dim(filter(counts_tops, top_in == "txrevise"))[1] # 0.3%
dim(filter(counts_tops, TPM == 0, top_in == "Both"))[1] / dim(filter(counts_tops, top_in == "Both"))[1] # <0.1%

medians = counts_tops %>%
  group_by(top_in) %>%
  summarise(top_in_median=median(TPM))
counts_tops = inner_join(counts_tops, medians)

ggplot(counts_tops, aes(x=TPM, y=..count.., fill=top_in, colour=top_in)) +
  geom_density(alpha=0.2) +
  scale_x_log10(limits=c(0.01, 10000), breaks=c(0.01, 0.1, 1, 10, 100, 1000, 10000)) + 
  labs(y="Number of gene ocurrences", fill="Significant in", color="Significant in") + 
  geom_vline(data=medians, aes(xintercept=top_in_median, color=top_in), size=1.5, alpha=0.5)

txrevise_log_TPMs = counts_tops %>%
  filter(top_in == "txrevise", TPM > 0) %>%
  mutate(logTPM = log10(TPM)) %>%
  select(logTPM)
cage_log_TPMs = counts_tops %>%
  filter(top_in == "CAGE", TPM > 0) %>%
  mutate(logTPM = log10(TPM)) %>%
  select(logTPM)

t.test(cage_log_TPMs$logTPM, txrevise_log_TPMs$logTPM, alternative="less") # p-value < 2.2e-16
# Nonparametric test - Mann-Whitney test / Wilcoxon Rank Sum test
wilcox.test(cage_log_TPMs$logTPM, txrevise_log_TPMs$logTPM, alternative="less") # p-value < 2.2e-16
```

# Apply expression filter

```{r}
threshold_tpm = 1
threshold_frac = 0.05

by_genes = counts_common %>%
  group_by(gene) %>%
  summarise(enough_frac = mean(TPM >= threshold_tpm),
            enough_tpm = quantile(TPM, probs=1-threshold_frac))

ggplot(by_genes, aes(enough_frac)) +
  geom_histogram()

non_zero_tpms = by_genes %>% filter(enough_tpm > 0)
non_zero_tpms = as.vector(non_zero_tpms$enough_tpm)

# fit two normal distributions
to_fit = as.vector(log10(non_zero_tpms))
model = normalmixEM(to_fit, lambda=c(0.6, 0.4), mu=c(-1, 1), sigma=c(1.5,0.5), k=2)

norm1 = function(x) {return( dnorm(x, mean=model$mu[1], sd=model$sigma[1]) / 2)}
norm2 = function(x) {return( dnorm(x, mean=model$mu[2], sd=model$sigma[2]) / 2)}
joined = function(x) {return( norm1(x) + norm2(x) )}

ggplot(mapping=aes(x=to_fit)) +
  geom_histogram(aes(y=..density.., group="hist"), alpha=0.6, bins=60) +
  scale_x_continuous(breaks=c(-4, -3, -2, -1, 0, 1, 2, 3, 4), limits=c(-4,4)) + 
  stat_function(fun=norm1, color="red") + 
  stat_function(fun=norm2, color="green") + 
  stat_function(fun=joined, color="black") +
  labs(x="Required minimum TPM, log10")

enough_genes = by_genes$gene[by_genes$enough_tpm >= threshold_tpm]
length(enough_genes)/length(beginning_common_genes) # 62.9% (31.5% for 10)
common_genes = enough_genes
```

Run "Filter results", "Analysis", "To visualize" and optionally "Optimizing annotations' N" again now

# Visualization

```{r}
results_table = tibble(`Min TPM`=c("0", "0.1", "1", "10", "100"), CAGE=c(1208, 1235, 1145, 564, 87), 
                       Txrevise=c(1073, 1051, 979, 663, 152), Augmented=c(1393, 1359, 1287, 868, 201))
results_table = melt(results_table, id.vars=c("Min TPM"), variable.name="Type", value.name="Result")
ggplot(results_table, aes(x=`Min TPM`, y=Result, color=Type, group=Type)) + 
  geom_line() + 
  geom_point() +
  scale_x_discrete() + 
  ylim(0, 1500) + 
  labs(y="puQTLs found")
```

```{r}
added = setdiff(intersect(bwa_tops$gene, ann_tops$gene), tx25_tops$gene) # 135
removed = setdiff(intersect(bwa_tops$gene, tx25_tops$gene), ann_tops$gene) # 45

getgrouping = function(x) {
  if (x %in% added) {return("Became significant")}
  if (x %in% removed) {return("Lost significance")}
  return("No change")
}

comparison = merge(select(bwa_results, gene, p_adjusted_beta_adjusted),
                   select(txrevise25_results, gene, p_adjusted_beta_adjusted),
                   by="gene")
colnames(comparison) = c("gene", "bwa_p", "tx_p")
comparison$grouping = as.factor(sapply(comparison$gene, getgrouping))
comparison = arrange(comparison, grouping)

fig1 = ggplot(comparison, aes(x=-log10(tx_p), y=-log10(bwa_p), color=grouping)) +
  labs(x="txrevise -log10(p)", y="CAGE -log10(p)", color="Change") +
  geom_smooth(aes(x=-log10(tx_p), y=-log10(bwa_p)), inherit.aes=FALSE, method="lm")


comparison = merge(select(bwa_results, gene, p_adjusted_beta_adjusted),
                   select(annots_results, gene, p_adjusted_beta_adjusted),
                   by="gene")
colnames(comparison) = c("gene", "bwa_p", "ann_p")
comparison$grouping = as.factor(sapply(comparison$gene, getgrouping))
comparison = arrange(comparison, grouping)

fig2 = ggplot(comparison, aes(x=-log10(ann_p), y=-log10(bwa_p), color=grouping)) +
  labs(x="Supplemented txrevise -log10(p)", y="CAGE -log10(p)", color="Change") +
  geom_smooth(aes(x=-log10(ann_p), y=-log10(bwa_p)), inherit.aes=FALSE, method="lm")


comparison = merge(select(annots_results, gene, p_adjusted_beta_adjusted),
                   select(txrevise25_results, gene, p_adjusted_beta_adjusted),
                   by="gene")
colnames(comparison) = c("gene", "ann_p", "tx_p")
comparison$grouping = as.factor(sapply(comparison$gene, getgrouping))
comparison = arrange(comparison, grouping)

fig3 = ggplot(comparison, aes(x=-log10(tx_p), y=-log10(ann_p), color=grouping)) +
  labs(x="txrevise -log10(p)", y="Supplemented txrevise -log10(p)", color="Change") +
  geom_smooth(aes(x=-log10(tx_p), y=-log10(ann_p)), inherit.aes=FALSE, method="lm")


for (fig in list(fig1, fig2, fig3)) {
  print(fig +
          geom_segment(x=0, y=0, xend=120, yend=120, color="black") +
          geom_point(alpha=0.4) +
          geom_point(aes(color=grouping), alpha=0.4) +
          scale_color_manual(values=c("green", "red", "dark gray")) +
          xlim(0,50) + ylim(0,50) +
          theme(legend.position="top"))
}
```

```{r}
sample_mapping = read_tsv("../qtlmap_prep/sampleMetadata.tsv") %>%
  select(sample_id, genotype_id) %>%
  mutate(source="cage")
sample_mapping_ = read_tsv("GEUVADIS_EUR.tsv") %>%
  select(sample_id, genotype_id) %>%
  mutate(source="txrevise")
sample_mapping = rbind(sample_mapping, sample_mapping_)

genotypes = read_tsv("variantinfo.vcf")[c(3,10:163)] %>%
  melt(id.vars=c("ID")) %>%
  mutate(source="cage")
genotypes_ = read_tsv("variantinfo_geuvadis.vcf")[c(3,10:454)] %>%
  melt(id.vars=c("ID")) %>%
  mutate(source="txrevise")
genotypes = rbind(genotypes, genotypes_)

names(genotypes) = c("variant", "genotype_id", "alleles", "source")
genotypes = genotypes %>%
  merge(sample_mapping) %>%
  select(-genotype_id, -source)

mapping = list(0, 1, 1, 2)
names(mapping) = c("0|0", "0|1", "1|0", "1|1")
genotypes$alleles = mapping[genotypes$alleles]

counts_bwa = read_tsv("../qtlmap_prep/countMatrix.tsv") %>%
  melt(id.vars=c("phenotype_id"), variable.name=c("sample_id"), value.name="proportion")
#counts_tx = read_tsv("/countMatrix_txrevise.tsv") %>% # 2.3G raw
#  melt(id.vars=c("phenotype_id"), variable.name=c("sample_id"), value.name="proportion")

#dataset, counts file, promoter, variant
for (info in list(
  list(bwa_tops, 1, counts_bwa),
  list(bwa_tops, 2, counts_bwa)
)) {
  dataset = info[[1]]
  n = info[[2]]
  counts = info[[3]]

  this_promoter = dataset$top_phenotype[n]
  this_variant = dataset$top_variant[n]

  counts_subset = filter(counts, phenotype_id==this_promoter) %>%
    select(sample_id, proportion)
  genotypes_subset = filter(genotypes, variant==this_variant) %>%
    select(sample_id, alleles)
  to_plot = merge(counts_subset, genotypes_subset)
  to_plot$alleles = as.numeric(to_plot$alleles)

  coef = coef(lm(proportion ~ alleles, data=to_plot))
  p = ggplot(to_plot, aes(x=alleles, y=proportion, group=as.factor(alleles))) +
        geom_jitter(width=0.1, color="black", alpha=0.7) +
        geom_boxplot(alpha=0, lwd=1, color="black") +
        scale_x_continuous(breaks=c(0,1,2)) +
        geom_abline(intercept=coef[1], slope=coef[2], lwd=1) +
        labs(x="Aternatiivsete alleelide arv", y="Promootori normaliseeritud osakaalud")
  print(p)
}
```
