---
title: "Convert"
subtitle: Andreas Vija
output:
  pdf_document: default
  pdf: default
  html_document:
    df_print: paged
editor_options:
  chunk_output_type: console
---

```{r setup, include=FALSE}
# This notebook uses featureCounts output and promoter annotations to build a count matrix
# and converts promoter annotations and sample metadata to the format qtlmap wants

knitr::opts_chunk$set(echo = TRUE, message = FALSE, warning = FALSE)
setwd("~/cage/qtlmap_prep")
```

```{r}
library("GenomicFeatures")
library("dplyr")
library("ggplot2")
library("readr")
library("stringr")
library("reshape2")
library("tidyr")
set.seed(123)
```


```{r}
summ = read_tsv("bwa.counts.summary", col_types = paste0(c("c", rep("n", 154)), collapse=""))
summ = unname(summ[c(1,2,10),])
summ = t(summ)[2:155,]

assigned = as.numeric(summ[,1])
unmapped = as.numeric(summ[,2])
nofeatures = as.numeric(summ[,3])

ggplot() + geom_histogram(aes(x=assigned, fill="assigned")) +
  geom_histogram(aes(x=unmapped, fill="unmapped")) +
  geom_histogram(aes(x=nofeatures, fill="nofeatures"))

assigned_total = sum(assigned)
unmapped_total = sum(unmapped)
nofeatures_total = sum(nofeatures)
total = assigned_total + unmapped_total + nofeatures_total

assigned_total / total #  34.0%
unmapped_total / total #   4.8%
nofeatures_total / total #61.2%
```


```{r}
data = read_tsv("bwa.counts", skip=1, col_types = paste0(c("ccnncn", rep("n", 154)), collapse=""))
data = select(data, -"Chr", -"Start", -"End", -"Strand", -"Length")

renamer <- function(name) {
  name = str_replace(name, "Geneid", "promoter")
  name = str_replace(name, "../align/results/", "")
  name = str_replace(name, ".sorted.bam", "")
  return(name)
}
colnames(data) = unname(sapply(colnames(data), renamer))
```


```{r}
genepromoters = read_tsv("FANTOM5_promoter_annotations.tsv", col_names=c("promoter", "gene"), skip=1)

# add gene name in addition to promoter
data = inner_join(genepromoters, data, by="promoter")
remove(genepromoters)

data = melt(data, id.vars=c("gene", "promoter"), variable.name="sample", value.name="count")
data$sample = as.character(data$sample)
```

```{r}
geneCountsInPerson = data %>%
  group_by(gene, sample) %>%
  summarise(geneCount=sum(count)) %>%
  ungroup()

geneCounts = geneCountsInPerson %>%
  group_by(gene) %>%
  summarise(geneCount=sum(geneCount)) %>%
  ungroup()

'
summary(geneCountsInPerson)

ggplot(data=geneCountsInPerson, aes(x=geneCount+1e-1)) + #+1e-1 allows to see zeroes
  geom_histogram() +
  scale_x_log10() +
  labs(title="Promoter reads per gene in one person")

ggplot(data=geneCounts, aes(x=geneCount+1e-1)) +
  geom_histogram() +
  scale_x_log10() +
  labs(title="Promoter reads per gene over all people")
'
```

```{r}
# filter out non-detected genes
lower_threshold = 0
commonGenes = (filter(geneCounts, geneCount > lower_threshold))$gene # 946 / 20 193 are zeroes (4.7%) (1594 promotors)
data = filter(data, gene %in% commonGenes) # 245 476 / 14 407 316 samples removed (1.7%)
bwa_kept_genes = unique(data$gene)

#calculate promoter proportions
data = inner_join(geneCountsInPerson, data, by=c("gene", "sample"))

data = data %>%
  mutate(promoterProportion = count/geneCount) %>%
  select(gene, sample, promoter, promoterProportion)

ggplot(mapping=aes(x=data$promoterProportion)) +
  geom_histogram() +
  labs(title="Proportions of promoters' importances")
```


```{r}
# create count matrix

data = data %>%
  mutate(phenotype_id=promoter, sample_id=sample) %>%
  select(phenotype_id, sample_id, promoterProportion) %>%
  unique()

countMatrix = data %>% dcast(phenotype_id ~ sample_id, sum, value.var="promoterProportion")
```


```{r}
phenotype_id = countMatrix[,1]
numerics = countMatrix[,2:155]

# Replace NAs with row means
na_indexes = which(is.na(numerics), arr.ind=TRUE)
numerics[na_indexes] = rowMeans(numerics, na.rm=TRUE)[na_indexes[,1]]

# inverse normal transformation
int <- function(arr) {
  return(qnorm( rank(arr, ties.method="random") / (length(arr)+1) )) # +1 because 100% quantile is infinity
}
numerics = as.data.frame(t(apply(numerics, 1, int)))

countMatrix = cbind(phenotype_id, numerics)
```


```{r}
write.table(countMatrix, file="countMatrix.tsv", sep="\t", row.names=FALSE, quote=FALSE)
```


```{r}
# convert FANTOM5 promoter metadata into pipeline format (phenotype_id, group_id, gene_id, chromosome, phenotype_pos, strand)

# calculate placeholder phenotype_pos as promotor average position

promoterMetadata = read_tsv("FANTOM5_promoter_annotations.tsv", col_types="ccciiicii") %>%
  mutate(phenotype_id=tss_id, group_id=gene_name, gene_id=gene_name, chromosome=chr,
         promoter_pos=peak_start, strand=if_else(strand == "+", "1", "-1")) %>%
  select(phenotype_id, group_id, gene_id, chromosome, promoter_pos, strand)

genePromoterPositions = promoterMetadata %>%
  group_by(gene_id) %>%
  summarise(n=n(), avg_prom=mean(promoter_pos), furthest_prom=max(promoter_pos) - min(promoter_pos)) %>%
  ungroup()

ggplot(genePromoterPositions, aes(x=furthest_prom+1e-1)) +
  geom_histogram() +
  scale_x_log10() +
  labs(title="Every gene's promoter span")

sum(genePromoterPositions$furthest_prom > 200000)/length(genePromoterPositions$furthest_prom)
# 2.7% of genes won't even have all promoter locations captured in a 200kb search range

geneAvgs = promoterMetadata %>%
  group_by(gene_id) %>%
  summarise(phenotype_pos = as.integer(mean(promoter_pos))) %>%
  ungroup()

promoterMetadata = promoterMetadata %>%
  inner_join(geneAvgs, by='gene_id') %>%
  select(phenotype_id, group_id, gene_id, chromosome, phenotype_pos, strand)

# replace phenotype pos with the gene's location

promoterMeta = read_tsv("transcript_usage_Ensembl_96_phenotype_metadata.tsv", col_types="ccccciicccci") %>%
  select(gene_id, phenotype_pos) %>%
  unique() %>%
  mutate(gene_pos = phenotype_pos) %>%
  select(gene_id, gene_pos)

promoterMetadata = promoterMetadata %>% left_join(promoterMeta, by='gene_id')
promoterMetadata$phenotype_pos = promoterMetadata$gene_pos

promoterMetadata = promoterMetadata %>%
  select(phenotype_id, group_id, gene_id, chromosome, phenotype_pos, strand)
write.table(promoterMetadata, file="phenotypeMetadata.tsv", sep="\t", row.names=FALSE, quote=FALSE)
```


```{r}
# convert sample metadata into pipeline format (sample_id, genotype_id, qtl_group)

sampleMetadata = read_tsv("Garieri_sample_metadata.txt") %>%
  mutate(qtl_group="LCL") %>%
  select(sample_id, genotype_id, qtl_group)

write.table(sampleMetadata, file="sampleMetadata.tsv", sep="\t", row.names=FALSE, quote=FALSE)
```


```{r}
# create a list of shared genes

upstream1 = GenomicFeatures::makeTxDbFromGFF("../analysis/txrevise/scripts/processed/Homo_sapiens.GRCh38.96_regular/txrevise_regular.grp_1.upstream.gff3")
upstream2 = GenomicFeatures::makeTxDbFromGFF("../analysis/txrevise/scripts/processed/Homo_sapiens.GRCh38.96_regular/txrevise_regular.grp_2.upstream.gff3")
genes1 = AnnotationDbi::select(upstream1, keys=keys(upstream1, "GENEID"), keytype="GENEID", columns="GENEID")
genes2 = AnnotationDbi::select(upstream2, keys=keys(upstream2, "GENEID"), keytype="GENEID", columns="GENEID")
genes1 = unlist(genes1, use.names=FALSE)
genes2 = unlist(genes2, use.names=FALSE)

fun = function(x, n=1) {
  return(strsplit(x, ".", fixed=TRUE)[[1]][n])
}

genes = union(sapply(genes1, fun), sapply(genes2, fun))
common_genes = intersect(bwa_kept_genes, genes)
saveRDS(common_genes, "common_genes.rds")
```
